# Overfitting and underfitting

In the first lecture, there was a question about optimizing with respect to the training data.
Remember that one of the central goals of machine learning is to generalize to unseen data.
How are these two notions related?

Let's take a step back and reconsider the classification and regression task we have seen so far.
In both cases we are persented with a finite dataset consisting of samples and labels.
We might fix a performance measure, for example for $k$NN this was given by measuring the percentage of correct answers and for linear regression this was given by the mean squared error.

Then we split our dataset into training set and test set and train our model on the training set and train our model on the training set. In case of $k$NN, this was done by simply storing the data so that we have access to it.
In case of linear regression, training consists of finding the minimum of the loss/cost function on the training set.
This results in a training error.

Then we measure performance on the test set which results in a test error.
Now the test error will in general be greater or equal to the training error and the goal is to find models so that the gap between them is small (and the errors are small).

Now our model is **overfitted**, if the training error is small but the test error is large.
If both errors are large then we call the model **underfitted**.

There are several possibilities to counter this.
* If our model has parameters, we might try and change them. $k$NN does have parameters (the number of neighbors), while linear regression does not have any.
* Change the hypothesis space, i.e. change the space of functions we consider. For example only consider polynomials up to a certain degree.
* Change the model.


# Example using $k$NN

The only parameter we can change in the $k$NN-algorithm is the number of neighbours.
Let's see how the number of neighbours affects how well we are able to predict an outcome or how well we are able to generalize.

What should be clear is that using only 1 neighbor will fit perfectly the training data.
However, we might not get optimal results for prediction.
This would be a classical case of overfitting (although this might still be good enough as the Iris flower example shows).
On the other hand, using larger number of neighbours somehow simplifies our model.
In the extreme case of using the same size as our dataset results in just taking the average of information available.
In this case the model would clearly be underfitted.
Let's use an example to illustrate this.

This time we use the Wisconsin Breast Cancer data set, which records clinical measurements of breast cancer tumors. Each tumor is labeled *benign* or *malignant* and the task is to predict whether a tumor is malignant based on the measurements of the tissue.

In [1]:
from sklearn.datasets import load_breast_cancer

cancer = load_breast_cancer()

print(cancer.keys())

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])


The `data` value contains the measurements, let's look at its shape.

In [2]:
print(cancer.data.shape)

(569, 30)


So we have 569 data points with 30 features each.
The features this time are:

In [3]:
print(cancer.feature_names)

['mean radius' 'mean texture' 'mean perimeter' 'mean area'
 'mean smoothness' 'mean compactness' 'mean concavity'
 'mean concave points' 'mean symmetry' 'mean fractal dimension'
 'radius error' 'texture error' 'perimeter error' 'area error'
 'smoothness error' 'compactness error' 'concavity error'
 'concave points error' 'symmetry error' 'fractal dimension error'
 'worst radius' 'worst texture' 'worst perimeter' 'worst area'
 'worst smoothness' 'worst compactness' 'worst concavity'
 'worst concave points' 'worst symmetry' 'worst fractal dimension']


You can read more about this dataset by printing out the `DESCR` value if you like.
Now let's see how the $k$NN algorithm performs using this dataset.
We first split into training and testing.

In [4]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(cancer.data, cancer.target, 
                                                   stratify=cancer.target, random_state=66)

Now we want to plot how well the the algorithm performs with respect to the number of neighbors.

In [7]:
training_accuracy = []
test_accuracy = []

neighbors_settings = range(1,560)

from sklearn.neighbors import KNeighborsClassifier

for k in neighbors_settings:
    clf = KNeighborsClassifier(n_neighbors=k)
    clf.fit(X_train, y_train)
    training_accuracy.append(clf.score(X_train, y_train))
    test_accuracy.append(clf.score(X_test, y_test))


Let's plot this.

In [ ]:
%matplotlib notebook

import matplotlib.pyplot as plt

plt.plot(neighbors_settings, training_accuracy, label='training accuracy')
plt.plot(neighbors_settings, test_accuracy)